internal/auth/providers/groups_cache.go

```diff
@@ -0,0 +1,59 @@
+package providers
+
+import (
+	"reflect"
+	"time"
+
+	"github.com/buzzfeed/sso/internal/pkg/groups"
+	"github.com/datadog/datadog-go/statsd"
+)
+
+type Cache interface {
+	Get(keys string)
+	Set(entries []string)
+	Purge(keys []string)
+}
+
+type GroupCache struct {
```

Comment:
This struct isn't fulfilling the interface, which I think we talked about this doing. 

Let's add a compile time check

```golang
// compile time interface check
var _ Provider =&GroupCache{}
```

And we can then fulfill the interface in one of two ways:

### 1. Embed the provider.
We can embed the provider in the struct by doing the following:

```golang
type GroupCache struct {
    metrics *statsd.Client
    cache *groups.LocalCache
    
    Provider
}
```

### 2. Stub out the rest of the interface
We supply stub methods for each of the interface methods which simply just call the underlying provider
```golang
func (p* GroupCache) OtherInterfaceFuncs( . . .{
    return p.provider.OtherInterFaceFuncs(...
}
```

Generally, 2 is preferred as it is more explicit and highlights the complexities of changing the interface if necessary. 1 is a convenient and can be useful if the changes you are making are minimal.

Reply:
Thank you for the explanations + examples!